File: benchmarks/weird_captcha_gym/shared_runtime/app/mechanics/three_camera_claw_machine.js
(() => {
  "use strict";
  let model = null;
  const clean = (value) =>
    String(value ?? "").replaceAll("&", "&amp;").replaceAll("<", "&lt;")
      .replaceAll(">", "&gt;").replaceAll('"', "&quot;");
  const round = (value) => Math.round(Number(value) * 10000) / 10000;
  function record(kind, details = {}) {
    const event = { sequence: model.events.length + 1, kind, ...details };
    model.events.push(event);
    return event;
  }
  function cloneObjects(objects = model.objects) {
    return Object.fromEntries(
      Object.entries(objects).map((
        [id, obj],
      ) => [id, { ...obj, center: [...obj.center] }]),
    );
  }
  function project(camera, point) {
    return camera.matrix.map((row, index) =>
      round(
        camera.origin[index] +
          row.reduce((sum, value, column) => sum + value * point[column], 0),
      )
    );
  }
  function digest(camera, claw, objects, captured) {
    const items = [
      ["claw", claw],
      ...Object.entries(objects).sort().map((
        [id, obj],
      ) => [id, id === captured ? claw : obj.center]),
    ];
    return items.map(([id, point]) => {
      const p = project(camera, point);
      return `${id}:${p[0].toFixed(4)}:${p[1].toFixed(4)}`;
    }).join("|");
  }
  function pointAabb(point, obstacle, radius) {
    return point.every((value, index) =>
      Math.abs(value - obstacle.center[index]) <= obstacle.half[index] + radius
    );
  }
  function blocker(start, end, radius) {
    const distance = Math.hypot(...end.map((v, i) => v - start[i])),
      steps = Math.max(
        1,
        Math.ceil(distance / model.state.world.collision_step),
      ),
      bounds = model.state.world.bounds;
    for (let n = 1; n <= steps; n++) {
      const t = n / steps, p = start.map((v, i) => v + (end[i] - v) * t);
      for (const [i, axis] of ["x", "y", "z"].entries()) {
        if (
          p[i] - radius < bounds[axis][0] || p[i] + radius > bounds[axis][1]
        ) return "cage-boundary";
      }
      for (const obstacle of model.state.obstacles) {
        if (pointAabb(p, obstacle, radius)) return obstacle.id;
      }
    }
    return null;
  }
  function contained(point, radius, box) {
    return point.every((v, i) =>
      Math.abs(v - box.center[i]) + radius <= box.half[i] + 1e-6
    );
  }
  function snapshot() {
    return {
      position: [...model.position],
      objects: cloneObjects(),
      captured: model.captured,
    };
  }
  function frameState(viewId) {
    const camera = model.state.cameras[viewId],
      visibleTick = Math.max(0, model.tick - camera.delay),
      snap = model.history[visibleTick];
    return {
      tick: visibleTick,
      digest: digest(camera, snap.position, snap.objects, snap.captured),
      snapshot: snap,
    };
  }
  function updateHistory() {
    if (model.captured) {
      model.objects[model.captured].center = [...model.position];
    }
    model.history[model.tick] = snapshot();
  }
  function clearFresh() {
    const root = document.querySelector(".claw-captcha"),
      wasFresh = root?.dataset.freshFailure === "true";
    document.querySelector(".claw-fresh")?.remove();
    root?.removeAttribute("data-fresh-failure");
    if (wasFresh && model && !model.terminal && !model.delivered) {
      setMessage("FRESH CAGE ACTIVE · TRIANGULATE THE MARKED ARTIFACT");
    }
  }
  function setMessage(message, status = "idle") {
    model.helpers.setReadout(message, status);
  }
  function applyControl(axis, direction = 0) {
    if (model.terminal || model.submitting || model.delivered) return;
    clearFresh();
    const world = model.state.world;
    if (axis === "brake") model.velocity = model.velocity.map((v) => v * .15);
    else if (axis === "coast") {}
    else {
      const index = { x: 0, y: 1, z: 2 }[axis];
      model.velocity[index] += direction * world.acceleration;
      const speed = Math.hypot(...model.velocity);
      if (speed > world.max_speed) {
        model.velocity = model.velocity.map((v) => v * world.max_speed / speed);
      }
    }
    record("control", {
      axis,
      ...(axis === "x" || axis === "y" || axis === "z" ? { direction } : {}),
    });
    physicsTick();
  }
  function physicsTick() {
    const start = [...model.position],
      candidate = model.position.map((v, i) => v + model.velocity[i]),
      radius = model.captured
        ? Math.max(
          model.state.world.claw_radius,
          model.objects[model.captured].radius,
        )
        : model.state.world.claw_radius,
      contact = blocker(start, candidate, radius);
    let resolution = "full", resolved = candidate;
    if (contact) {
      resolved = [...start];
      let moved = false;
      for (let index = 0; index < 3; index++) {
        const trial = [...resolved];
        trial[index] = candidate[index];
        if (!blocker(resolved, trial, radius)) {
          resolved = trial;
          moved ||= Math.abs(trial[index] - start[index]) > 1e-9;
        } else model.velocity[index] = 0;
      }
      resolution = moved ? "slide" : "blocked";
      model.collisions++;
      const alert = document.querySelector(".claw-impact");
      alert.dataset.visible = "true";
      alert.querySelector("span").textContent =
        `${contact.toUpperCase()} · ${resolution.toUpperCase()}`;
      setMessage(
        `RIGID CONTACT · ${contact.toUpperCase()} · RECOVER WITH BRAKE / REVERSE`,
        "error",
      );
    } else document.querySelector(".claw-impact").dataset.visible = "false";
    model.position = resolved;
    model.velocity = model.velocity.map((v) => v * model.state.world.damping);
    model.tick++;
    updateHistory();
    const visibleFrames = {};
    Object.keys(model.state.cameras).forEach((view) => {
      const frame = frameState(view);
      visibleFrames[view] = { tick: frame.tick, digest: frame.digest };
    });
    record("physics_tick", {
      tick: model.tick,
      resolution,
      ...(contact ? { contact } : {}),
      position: model.position.map(round),
      velocity: model.velocity.map(round),
      visible_frames: visibleFrames,
    });
    drawFeeds();
    updateHUD();
  }
  function grip() {
    if (model.terminal || model.submitting || model.delivered) return;
    clearFresh();
    if (model.gripper === "open") {
      const candidates = Object.entries(model.objects).map((
        [id, obj],
      ) => [
        Math.hypot(...obj.center.map((v, i) => v - model.position[i])),
        id,
      ]).sort((a, b) => a[0] - b[0]);
      let captured = null;
      for (const [distance, id] of candidates) {
        const obj = model.objects[id];
        if (
          distance <= model.state.world.capture_distance &&
          !blocker(obj.center, model.position, obj.radius)
        ) {
          captured = id;
          break;
        }
      }
      model.gripper = "closed";
      model.captured = captured;
      record("gripper", { action: "close", captured_id: captured });
      if (captured) {
        model.objects[captured].center = [...model.position];
        setMessage(
          "GRIPPER LOAD DETECTED · LIFT ABOVE THE CAGE BAFFLES",
          "pending",
        );
      } else setMessage("GRIPPER CLOSED ON EMPTY AIR", "error");
    } else {
      const released = model.captured,
        delivered = Boolean(
          released && model.objects[released].marked &&
            contained(
              model.position,
              model.objects[released].radius,
              model.state.chute,
            ),
        );
      record("gripper", { action: "open", released_id: released, delivered });
      model.gripper = "open";
      model.captured = null;
      if (delivered) {
        model.delivered = model.terminal = true;
        document.querySelector(".claw-complete").dataset.visible = "true";
        setMessage("MARKED ARTIFACT CONTAINED IN DELIVERY CHUTE", "passed");
      } else {setMessage(
          released
            ? "LOAD RELEASED OUTSIDE CERTIFIED CHUTE"
            : "EMPTY GRIPPER OPENED",
          released ? "error" : "idle",
        );}
    }
    updateHistory();
    drawFeeds();
    updateHUD();
  }
  function resetClaw() {
    if (model.captured || model.delivered || model.submitting) return;
    record("reset_claw");
    model.resets++;
    model.position = [...model.state.initial.position];
    model.velocity = [...model.state.initial.velocity];
    model.objects = Object.fromEntries(
      model.state.objects.map(
        (obj) => [obj.id, { ...obj, center: [...obj.center] }],
      ),
    );
    model.gripper = "open";
    model.captured = null;
    model.tick = 0;
    model.history = { 0: snapshot() };
    document.querySelector(".claw-impact").dataset.visible = "false";
    drawFeeds();
    updateHUD();
    setMessage("CLAW AND CCTV BUFFERS REWOUND");
  }
  function drawFeed(viewId) {
    const canvas = document.querySelector(`.claw-feed[data-view='${viewId}']`),
      ctx = canvas.getContext("2d"),
      camera = model.state.cameras[viewId],
      frame = frameState(viewId),
      snap = frame.snapshot;
    ctx.clearRect(0, 0, canvas.width, canvas.height);
    ctx.fillStyle = "#0a100e";
    ctx.fillRect(0, 0, canvas.width, canvas.height);
    ctx.strokeStyle = "#90ad9640";
    ctx.lineWidth = 1;
    for (let i = 0; i < 12; i++) {
      ctx.beginPath();
      ctx.moveTo(i * canvas.width / 11, 0);
      ctx.lineTo(i * canvas.width / 11, canvas.height);
      ctx.stroke();
      ctx.beginPath();
      ctx.moveTo(0, i * canvas.height / 11);
      ctx.lineTo(canvas.width, i * canvas.height / 11);
      ctx.stroke();
    }
    function drawBox(center, half, color) {
      const p = project(camera, center),
        hx = Math.max(
          5,
          Math.abs(camera.matrix[0][0]) * half[0] +
            Math.abs(camera.matrix[0][1]) * half[1] +
            Math.abs(camera.matrix[0][2]) * half[2],
        ),
        hy = Math.max(
          5,
          Math.abs(camera.matrix[1][0]) * half[0] +
            Math.abs(camera.matrix[1][1]) * half[1] +
            Math.abs(camera.matrix[1][2]) * half[2],
        );
      ctx.fillStyle = color;
      ctx.strokeStyle = color;
      ctx.fillRect(p[0] - hx, p[1] - hy, hx * 2, hy * 2);
      ctx.strokeRect(p[0] - hx, p[1] - hy, hx * 2, hy * 2);
    }
    const scale = Math.max(...camera.matrix.flat().map(Math.abs));
    const primitives = [
      ...model.state.obstacles.map((obstacle) => ({
        kind: "box",
        center: obstacle.center,
        half: obstacle.half,
        color: "#303a35",
        priority: 2,
      })),
      {
        kind: "box",
        center: model.state.chute.center,
        half: model.state.chute.half,
        color: model.state.palette.chute,
        priority: 1,
      },
      ...Object.entries(snap.objects).map(([id, obj]) => ({
        kind: "object",
        center: id === snap.captured ? snap.position : obj.center,
        obj,
        priority: 3,
      })),
      { kind: "claw", center: snap.position, priority: 4 },
    ];
    primitives.sort((a, b) =>
      (a.center[camera.depth_axis] * camera.depth_sign) -
        (b.center[camera.depth_axis] * camera.depth_sign) ||
      a.priority - b.priority
    );
    primitives.forEach((primitive) => {
      if (primitive.kind === "box") {
        drawBox(primitive.center, primitive.half, primitive.color);
        return;
      }
      const p = project(camera, primitive.center);
      if (primitive.kind === "object") {
        const obj = primitive.obj;
        ctx.fillStyle = obj.color;
        ctx.strokeStyle = obj.marked ? "#fff" : "#18231e";
        ctx.lineWidth = obj.marked ? 3 : 1;
        ctx.beginPath();
        ctx.arc(p[0], p[1], Math.max(5, obj.radius * scale), 0, Math.PI * 2);
        ctx.fill();
        ctx.stroke();
        if (obj.marked) {
          const radius = obj.radius * scale;
          ctx.beginPath();
          ctx.moveTo(p[0], p[1] - radius - 4);
          ctx.lineTo(p[0] - 6, p[1] - radius - 12);
          ctx.lineTo(p[0] + 6, p[1] - radius - 12);
          ctx.closePath();
          ctx.fillStyle = "#fff";
          ctx.fill();
        }
        return;
      }
      const radius = model.state.world.claw_radius * scale;
      ctx.strokeStyle = model.state.palette.claw;
      ctx.fillStyle = model.state.palette.claw;
      ctx.lineWidth = 3;
      ctx.beginPath();
      ctx.arc(p[0], p[1], Math.max(5, radius), 0, Math.PI * 2);
      ctx.stroke();
      ctx.beginPath();
      ctx.moveTo(p[0] - radius - 4, p[1] + radius + 5);
      ctx.lineTo(p[0], p[1] + 3);
      ctx.lineTo(p[0] + radius + 4, p[1] + radius + 5);
      ctx.stroke();
    });
    ctx.fillStyle = "#d8eadc";
    ctx.font = "700 10px ui-monospace";
    ctx.fillText(
      `${camera.label} · FRAME T${frame.tick} · DELAY ${camera.delay}`,
      10,
      15,
    );
    document.querySelector(`.claw-feed-tick[data-view='${viewId}']`)
      .textContent = `T${frame.tick}`;
  }
  function drawFeeds() {
    Object.keys(model.state.cameras).forEach(drawFeed);
  }
  function updateHUD() {
    document.querySelector(".claw-speed").textContent = Math.hypot(
      ...model.velocity,
    ).toFixed(2);
    document.querySelector(".claw-grip-state").textContent = model.gripper
      .toUpperCase();
    document.querySelector(".claw-load-state").textContent = model.captured
      ? "LOAD"
      : "EMPTY";
    document.querySelector(".claw-tick").textContent = String(model.tick);
    document.querySelector(".claw-grip").textContent = model.gripper === "open"
      ? "CLOSE GRIPPER"
      : "OPEN / RELEASE";
  }
  async function submit() {
    if (model.submitting || model.terminal && !model.delivered) return;
    model.submitting = true;
    setMessage(
      "REPLAYING INERTIAL TICKS, SWEPT CAGE CONTACTS, AND STAGGERED FEEDS…",
      "pending",
    );
    const payload = {
      mechanic_id: model.state.mechanic_id,
      task_id: model.state.task_id,
      challenge_id: model.state.challenge_id,
      events: model.events,
      delivered: model.delivered,
      position: model.position.map(round),
      velocity: model.velocity.map(round),
      captured_id: model.captured,
      gripper: model.gripper,
      ticks: model.tick,
      collisions: model.collisions,
      resets: model.resets,
      feeds_seen: Object.keys(model.state.cameras).sort(),
    };
    try {
      const response = await fetch("/result", {
          method: "POST",
          headers: { "content-type": "application/json" },
          body: JSON.stringify(payload),
        }),
        outcome = await response.json();
      if (outcome.passed === true) {
        model.terminal = true;
        document.querySelector(".claw-captcha").insertAdjacentHTML(
          "beforeend",
          '<div class="claw-verdict"><small>STAGGERED CCTV CHAIN VERIFIED</small><strong>PASS</strong><span>INERTIA · MULTI-VIEW PROJECTION · GEOMETRIC DELIVERY</span></div>',
        );
        setMessage("PASS", "passed");
      } else if (outcome.passed === false && outcome.state) {
        await model.helpers.render(outcome.state);
        const root = document.querySelector(".claw-captcha");
        root.dataset.freshFailure = "true";
        root.insertAdjacentHTML(
          "afterbegin",
          '<div class="claw-fresh"><b>FAIL</b><span>RETRIEVAL REJECTED · FRESH CAGE ISSUED</span></div>',
        );
        const readout = document.querySelector(".readout");
        readout.textContent = "FAIL · FRESH THREE-CAMERA CAGE ISSUED";
        readout.dataset.status = "error";
        setTimeout(clearFresh, 1800);
      } else {
        model.submitting = false;
        setMessage("FAIL · NO AUTHORITATIVE CLAW GRADE", "error");
      }
    } catch (_) {
      model.submitting = false;
      setMessage("FAIL · CLAW VERIFIER OFFLINE", "error");
    }
  }
  async function render(state, helpers) {
    document.body.dataset.mechanic = "three-camera-claw-machine";
    helpers.app.innerHTML = `<section class="claw-captcha" data-challenge-id="${
      clean(state.challenge_id)
    }"><header class="claw-head"><div><span>REMOTE RECOVERY CELL / ${
      clean(state.challenge_id)
    }</span><h1>${
      clean(state.prompt)
    }</h1></div><p>NO DIRECT WINDOW<br><b>THREE DELAYED CAMERA CLOCKS</b></p></header><main class="claw-main"><section class="claw-feeds">${
      Object.entries(state.cameras).map(([id, camera]) =>
        `<article><header><b>${
          clean(camera.label)
        }</b><span class="claw-feed-tick" data-view="${id}">T0</span></header><canvas class="claw-feed" data-view="${id}" width="340" height="230"></canvas></article>`
      ).join("")
    }</section><aside class="claw-console"><span class="claw-kicker">SIX-AXIS INERTIAL GANTRY</span><h2>Correct what the cameras have not shown you yet.</h2><div class="claw-controls">${
      ["x", "y", "z"].map((axis) =>
        `<div><b>${axis.toUpperCase()} AXIS</b><button class="claw-control" data-axis="${axis}" data-direction="-1">${axis.toUpperCase()}−</button><button class="claw-control" data-axis="${axis}" data-direction="1">${axis.toUpperCase()}+</button></div>`
      ).join("")
    }</div><button class="claw-control claw-coast" data-axis="coast">COAST ONE PHYSICS TICK</button><button class="claw-control claw-brake" data-axis="brake">BRAKE / DAMP INERTIA</button><button class="claw-grip">CLOSE GRIPPER</button><button class="claw-reset">REWIND GANTRY</button><div class="claw-stats"><span>SPEED<b class="claw-speed">0.00</b></span><span>GRIP<b class="claw-grip-state">OPEN</b></span><span>LOAD<b class="claw-load-state">EMPTY</b></span><span>TICK<b class="claw-tick">0</b></span></div></aside><div class="claw-impact" data-visible="false"><b>CAGE CONTACT</b><span>RECOVER</span></div><div class="claw-complete" data-visible="false"><b>DELIVERY CONTAINED</b><span>MARKED LOAD IN CHUTE</span></div></main><footer class="claw-foot"><div><span>DELAY-AWARE PHYSICS REPLAY</span><div class="readout" data-status="idle">TRIANGULATE THE MARKED ARTIFACT</div></div><button class="claw-submit">${
      clean(state.submit_label)
    }</button></footer>${helpers.cheatPanelTemplate()}</section>`;
    model = {
      state,
      helpers,
      events: [],
      position: [...state.initial.position],
      velocity: [...state.initial.velocity],
      objects: Object.fromEntries(
        state.objects.map(
          (obj) => [obj.id, { ...obj, center: [...obj.center] }],
        ),
      ),
      gripper: "open",
      captured: null,
      delivered: false,
      terminal: false,
      submitting: false,
      tick: 0,
      collisions: 0,
      resets: 0,
      history: {},
    };
    model.history[0] = snapshot();
    window.threeCameraClawMachineModel = model;
    document.querySelectorAll(".claw-control").forEach((button) =>
      button.addEventListener(
        "click",
        () =>
          applyControl(
            button.dataset.axis,
            Number(button.dataset.direction || 0),
          ),
      )
    );
    document.querySelector(".claw-grip").addEventListener("click", grip);
    document.querySelector(".claw-reset").addEventListener("click", resetClaw);
    document.querySelector(".claw-submit").addEventListener("click", submit);
    helpers.installCheatPanel();
    drawFeeds();
    updateHUD();
  }
  window.WeirdCaptchaMechanics = window.WeirdCaptchaMechanics || {};
  window.WeirdCaptchaMechanics.three_camera_claw_machine = {
    rootSelector: ".claw-captcha",
    render,
  };
})();

File: benchmarks/weird_captcha_gym/shared_runtime/server/incubator_graders/three_camera_claw_machine.py
from __future__ import annotations

import math
from typing import Any


MECHANIC_ID = "three_camera_claw_machine"


def _round(value: float) -> float:
    return round(float(value), 4)


def _project(camera: dict[str, Any], point: list[float]) -> list[float]:
    matrix = camera["matrix"]
    return [_round(camera["origin"][row] + sum(matrix[row][col] * point[col] for col in range(3))) for row in range(2)]


def _digest(camera: dict[str, Any], claw: list[float], objects: dict[str, dict[str, Any]], captured: str | None) -> str:
    items = [("claw", claw)] + [(key, claw if key == captured else value["center"]) for key, value in sorted(objects.items())]
    return "|".join(f"{key}:{screen[0]:.4f}:{screen[1]:.4f}" for key, point in items for screen in [_project(camera, point)])


def _point_aabb(point: list[float], obstacle: dict[str, Any], radius: float) -> bool:
    return all(abs(point[i] - obstacle["center"][i]) <= obstacle["half"][i] + radius for i in range(3))


def _blocker(start: list[float], end: list[float], contract: dict[str, Any], carried_radius: float) -> str | None:
    distance = math.dist(start, end); steps = max(1, math.ceil(distance / contract["world"]["collision_step"])); bounds = contract["world"]["bounds"]
    for step in range(1, steps + 1):
        t = step / steps; point = [start[i] + (end[i] - start[i]) * t for i in range(3)]
        for i, axis in enumerate(("x", "y", "z")):
            if point[i] - carried_radius < bounds[axis][0] or point[i] + carried_radius > bounds[axis][1]: return "cage-boundary"
        for obstacle in contract["obstacles"]:
            if _point_aabb(point, obstacle, carried_radius): return str(obstacle["id"])
    return None


def _contained(point: list[float], radius: float, box: dict[str, Any]) -> bool:
    return all(abs(point[i] - box["center"][i]) + radius <= box["half"][i] + 1e-6 for i in range(3))


def _frames(tick: int, history: dict[int, tuple[list[float], dict[str, dict[str, Any]], str | None]], cameras: dict[str, Any]) -> dict[str, Any]:
    result = {}
    for view_id, camera in cameras.items():
        visible_tick = max(0, tick - int(camera["delay"])); claw, objects, captured = history[visible_tick]
        result[view_id] = {"tick": visible_tick, "digest": _digest(camera, claw, objects, captured)}
    return result


def _clone_objects(items: list[dict[str, Any]]) -> dict[str, dict[str, Any]]:
    return {item["id"]: {**item, "center": list(item["center"])} for item in items}


def grade(payload: dict[str, Any], ground_truth: dict[str, Any], public_state: dict[str, Any]) -> dict[str, Any]:
    challenge, task_id = str(ground_truth.get("challenge_id") or ""), str(ground_truth.get("task_id") or "")
    if payload.get("mechanic_id") != MECHANIC_ID or ground_truth.get("mechanic_id") != MECHANIC_ID: return {"graded": True, "passed": False, "feedback": "mechanic mismatch"}
    if not challenge or payload.get("challenge_id") != challenge or public_state.get("challenge_id") != challenge: return {"graded": True, "passed": False, "feedback": "stale challenge"}
    if not task_id or payload.get("task_id") != task_id: return {"graded": True, "passed": False, "feedback": "task identity mismatch"}
    for field in ("task_id", "world", "initial", "objects", "obstacles", "chute", "cameras", "requirements"):
        if public_state.get(field) != ground_truth.get(field): return {"graded": True, "passed": False, "feedback": f"public/private claw {field} contract skew"}
    events = payload.get("events"); requirements = ground_truth["requirements"]
    if not isinstance(events, list) or not (1 <= len(events) <= requirements["max_events"]): return {"graded": True, "passed": False, "feedback": "claw transcript missing or outside limits"}
    initial = ground_truth["initial"]; position, velocity = list(initial["position"]), list(initial["velocity"]); objects = _clone_objects(ground_truth["objects"])
    gripper = "open"; captured = None; delivered = terminal = False; tick = collisions = resets = 0; pending_control = False; feeds_seen: set[str] = set()
    history = {0: (list(position), _clone_objects(list(objects.values())), captured)}
    for sequence, event in enumerate(events, 1):
        if terminal: return {"graded": True, "passed": False, "feedback": "interaction continued after chute delivery"}
        if not isinstance(event, dict) or event.get("sequence") != sequence: return {"graded": True, "passed": False, "feedback": f"event {sequence} sequence mismatch"}
        kind = event.get("kind")
        if kind == "control":
            if pending_control: return {"graded": True, "passed": False, "feedback": "control omitted its fixed physics tick"}
            axis = event.get("axis"); direction = int(event.get("direction", 0)); world = ground_truth["world"]
            if axis == "brake": velocity = [v * .15 for v in velocity]
            elif axis == "coast": pass
            elif axis in {"x", "y", "z"} and direction in {-1, 1}:
                index = {"x": 0, "y": 1, "z": 2}[axis]; velocity[index] += direction * world["acceleration"]
                speed = math.sqrt(sum(v * v for v in velocity))
                if speed > world["max_speed"]: velocity = [v * world["max_speed"] / speed for v in velocity]
            else: return {"graded": True, "passed": False, "feedback": "invalid six-axis claw control"}
            pending_control = True
        elif kind == "physics_tick":
            if not pending_control or event.get("tick") != tick + 1 or tick >= requirements["max_ticks"]: return {"graded": True, "passed": False, "feedback": "missing, duplicate, or excessive fixed physics tick"}
            tick += 1; start = list(position); candidate = [position[i] + velocity[i] for i in range(3)]
            carried_radius = ground_truth["world"]["claw_radius"] if captured is None else max(ground_truth["world"]["claw_radius"], objects[captured]["radius"])
            contact = _blocker(start, candidate, ground_truth, carried_radius); resolution = "full"
            if contact:
                resolved = list(start); moved = False
                for index in range(3):
                    trial = list(resolved); trial[index] = candidate[index]
                    if _blocker(resolved, trial, ground_truth, carried_radius) is None: resolved = trial; moved |= abs(trial[index] - start[index]) > 1e-9
                    else: velocity[index] = 0
                candidate = resolved; resolution = "slide" if moved else "blocked"; collisions += 1
            position = candidate; velocity = [v * ground_truth["world"]["damping"] for v in velocity]
            if captured: objects[captured]["center"] = list(position)
            history[tick] = (list(position), _clone_objects(list(objects.values())), captured); expected_frames = _frames(tick, history, ground_truth["cameras"])
            if event.get("resolution") != resolution or (str(event.get("contact") or "") if contact else None) != contact: return {"graded": True, "passed": False, "feedback": "claw collision/wall-slide claim disagrees with replay"}
            if not isinstance(event.get("position"), list) or math.dist([float(v) for v in event["position"]], position) > .01 or not isinstance(event.get("velocity"), list) or math.dist([float(v) for v in event["velocity"]], velocity) > .01: return {"graded": True, "passed": False, "feedback": "claw position or inertial velocity teleport"}
            if event.get("visible_frames") != expected_frames: return {"graded": True, "passed": False, "feedback": "staggered CCTV frame tick/projection digest drift"}
            feeds_seen.update(expected_frames); pending_control = False
        elif kind == "gripper":
            if pending_control: return {"graded": True, "passed": False, "feedback": "gripper operated between control and physics tick"}
            action = event.get("action")
            if action == "close" and gripper == "open":
                candidates = sorted((math.dist(position, obj["center"]), object_id) for object_id, obj in objects.items() if object_id != captured)
                expected = None
                for distance, object_id in candidates:
                    obj = objects[object_id]
                    if distance <= ground_truth["world"]["capture_distance"] and _blocker(obj["center"], position, ground_truth, obj["radius"]) is None:
                        expected = object_id; break
                if event.get("captured_id") != expected: return {"graded": True, "passed": False, "feedback": "fake geometric claw grab"}
                gripper = "closed"; captured = expected
                if captured: objects[captured]["center"] = list(position)
                history[tick] = (list(position), _clone_objects(list(objects.values())), captured)
            elif action == "open" and gripper == "closed":
                released = captured; expected_delivered = bool(released == ground_truth["target_id"] and _contained(position, objects[released]["radius"], ground_truth["chute"])) if released else False
                if event.get("released_id") != released or (event.get("delivered") is True) != expected_delivered: return {"graded": True, "passed": False, "feedback": "release/chute containment claim disagrees with replay"}
                gripper = "open"; captured = None
                history[tick] = (list(position), _clone_objects(list(objects.values())), captured)
                if expected_delivered: delivered = terminal = True
            else: return {"graded": True, "passed": False, "feedback": "invalid gripper transition"}
        elif kind == "reset_claw":
            if pending_control or captured: return {"graded": True, "passed": False, "feedback": "claw reset during motion primitive or carrying"}
            position, velocity = list(initial["position"]), list(initial["velocity"]); objects = _clone_objects(ground_truth["objects"]); gripper = "open"; tick = 0; history = {0: (list(position), _clone_objects(list(objects.values())), None)}; resets += 1
        else: return {"graded": True, "passed": False, "feedback": f"event {sequence} has unknown kind"}
    if pending_control: return {"graded": True, "passed": False, "feedback": "terminal transcript ended before physics tick"}
    summary = {"delivered": delivered, "position": [_round(v) for v in position], "velocity": [_round(v) for v in velocity], "captured_id": captured, "gripper": gripper, "ticks": tick, "collisions": collisions, "resets": resets, "feeds_seen": sorted(feeds_seen)}
    for field, value in summary.items():
        if field in {"position", "velocity"}:
            actual = payload.get(field)
            if not isinstance(actual, list) or len(actual) != 3 or math.dist([float(v) for v in actual], value) > .01: return {"graded": True, "passed": False, "feedback": f"submitted {field} disagrees with claw replay"}
        elif payload.get(field) != value: return {"graded": True, "passed": False, "feedback": f"submitted {field} disagrees with claw replay"}
    passed = delivered and set(summary["feeds_seen"]) == set(requirements["required_feeds"])
    return {"graded": True, "passed": passed, "score": 100 if passed else 0, "feedback": f"claw replay: ticks {tick}; feeds {len(feeds_seen)}/3; collisions {collisions}; marked artifact {'delivered' if delivered else 'inside'}"}


def cheat(public_state: dict[str, Any], ground_truth: dict[str, Any]) -> dict[str, Any]:
    del public_state
    return {"target_id": ground_truth.get("target_id"), "solver": ground_truth.get("solver") or {}}
